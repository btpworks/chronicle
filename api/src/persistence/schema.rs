// @generated automatically by Diesel CLI.

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    activity (id) {
        id -> Integer,
        name -> Text,
        namespace_id -> Integer,
        domaintype -> Nullable<Text>,
        started -> Nullable<Timestamp>,
        ended -> Nullable<Timestamp>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    activity_attribute (activity_id, typename) {
        activity_id -> Integer,
        typename -> Text,
        value -> Text,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    agent (id) {
        id -> Integer,
        name -> Text,
        namespace_id -> Integer,
        domaintype -> Nullable<Text>,
        current -> Integer,
        identity_id -> Nullable<Integer>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    agent_attribute (agent_id, typename) {
        agent_id -> Integer,
        typename -> Text,
        value -> Text,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    association (agent_id, activity_id, role) {
        agent_id -> Integer,
        activity_id -> Integer,
        role -> Nullable<Text>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    attachment (id) {
        id -> Integer,
        namespace_id -> Integer,
        signature_time -> Timestamp,
        signature -> Text,
        signer_id -> Integer,
        locator -> Nullable<Text>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    delegation (delegate_id, responsible_id, activity_id, role) {
        delegate_id -> Integer,
        responsible_id -> Integer,
        activity_id -> Nullable<Integer>,
        role -> Nullable<Text>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    derivation (activity_id, generated_entity_id, used_entity_id, typ) {
        activity_id -> Nullable<Integer>,
        generated_entity_id -> Integer,
        used_entity_id -> Integer,
        typ -> Nullable<Integer>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    entity (id) {
        id -> Integer,
        name -> Text,
        namespace_id -> Integer,
        domaintype -> Nullable<Text>,
        attachment_id -> Nullable<Integer>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    entity_attribute (entity_id, typename) {
        entity_id -> Integer,
        typename -> Text,
        value -> Text,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    generation (activity_id, generated_entity_id) {
        activity_id -> Integer,
        generated_entity_id -> Integer,
        typ -> Nullable<Text>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    hadattachment (entity_id, attachment_id) {
        entity_id -> Integer,
        attachment_id -> Integer,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    hadidentity (agent_id, identity_id) {
        agent_id -> Integer,
        identity_id -> Integer,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    identity (id) {
        id -> Integer,
        namespace_id -> Integer,
        public_key -> Text,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    ledgersync (correlation_id) {
        correlation_id -> Text,
        offset -> Nullable<Text>,
        sync_time -> Nullable<Timestamp>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    namespace (id) {
        id -> Integer,
        name -> Text,
        uuid -> Text,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    usage (activity_id, entity_id) {
        activity_id -> Integer,
        entity_id -> Integer,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    wasinformedby (namespace_id, activity_id, informing_activity_id) {
        namespace_id -> Integer,
        activity_id -> Integer,
        informing_activity_id -> Integer,
    }
}

diesel::joinable!(activity -> namespace (namespace_id));
diesel::joinable!(activity_attribute -> activity (activity_id));
diesel::joinable!(agent -> identity (identity_id));
diesel::joinable!(agent -> namespace (namespace_id));
diesel::joinable!(agent_attribute -> agent (agent_id));
diesel::joinable!(association -> activity (activity_id));
diesel::joinable!(association -> agent (agent_id));
diesel::joinable!(attachment -> identity (signer_id));
diesel::joinable!(attachment -> namespace (namespace_id));
diesel::joinable!(delegation -> activity (activity_id));
diesel::joinable!(derivation -> activity (activity_id));
diesel::joinable!(entity -> attachment (attachment_id));
diesel::joinable!(entity -> namespace (namespace_id));
diesel::joinable!(entity_attribute -> entity (entity_id));
diesel::joinable!(generation -> activity (activity_id));
diesel::joinable!(generation -> entity (generated_entity_id));
diesel::joinable!(hadattachment -> attachment (attachment_id));
diesel::joinable!(hadattachment -> entity (entity_id));
diesel::joinable!(hadidentity -> agent (agent_id));
diesel::joinable!(hadidentity -> identity (identity_id));
diesel::joinable!(identity -> namespace (namespace_id));
diesel::joinable!(usage -> activity (activity_id));
diesel::joinable!(usage -> entity (entity_id));
diesel::joinable!(wasinformedby -> activity (activity_id));
diesel::joinable!(wasinformedby -> namespace (namespace_id));

diesel::allow_tables_to_appear_in_same_query!(
    activity,
    activity_attribute,
    agent,
    agent_attribute,
    association,
    attachment,
    delegation,
    derivation,
    entity,
    entity_attribute,
    generation,
    hadattachment,
    hadidentity,
    identity,
    ledgersync,
    namespace,
    usage,
    wasinformedby,
);
