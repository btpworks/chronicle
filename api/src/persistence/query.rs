use super::schema::*;
use chrono::NaiveDateTime;
use diesel::prelude::*;

#[derive(Queryable)]
pub struct Namespace {
    pub name: String,
    pub uuid: String,
}

#[derive(Insertable)]
#[table_name = "namespace"]
pub struct NewNamespace<'a> {
    pub name: &'a str,
    pub uuid: &'a str,
}

#[derive(Insertable)]
#[table_name = "activity"]
pub struct NewActivity<'a> {
    pub name: &'a str,
    pub namespace: &'a str,
    pub started: Option<NaiveDateTime>,
    pub ended: Option<NaiveDateTime>,
}

#[derive(Queryable)]
pub struct Agent {
    pub id: i32,
    pub name: String,
    pub namespace: String,
    pub publickey: Option<String>,
    pub privatekeypath: Option<String>,
    pub current: i32,
}

#[derive(Queryable)]
pub struct Activity {
    pub id: i32,
    pub name: String,
    pub namespace: String,
    pub started: Option<NaiveDateTime>,
    pub ended: Option<NaiveDateTime>,
}

#[derive(Insertable, AsChangeset, Default)]
#[table_name = "agent"]
pub struct NewAgent<'a> {
    pub name: &'a str,
    pub namespace: &'a str,
    pub current: i32,
    pub publickey: Option<&'a str>,
    pub privatekeypath: Option<&'a str>,
}
