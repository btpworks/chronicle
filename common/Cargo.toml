[package]
build   = "build.rs"
edition = "2021"
name    = "common"
version = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
async-graphql = { features = [
  "opentelemetry",
  "chrono",
  "unblock",
  "default",
  "uuid",
], version = "3.0.38" }
async-trait = "0.1.53"
chrono = { version = "0.4.19", features = ["serde"] }
custom_error = "1.9.2"
derivative = "2.2.0"
diesel = { version = "2.0.0-rc.0", features = [
  "sqlite",
  "uuid",
  "chrono",
  "r2d2",
] }
futures = "0.3.21"
glob = "0.3.0"
hex = "0.4.3"
iref = "2.1.2"
iref-enum = "2.0.0"
json = "0.12.4"
json-ld = "0.4.0"
k256 = { version = "0.11.3", features = [
  "default",
  "arithmetic",
  "ecdsa",
  "pkcs8",
  "sha256",
  "std",
  "pem",
] }
lazy_static = "1.4.0"
openssl      = "0.10.38"
percent-encoding = "2.1.0"
pkcs8 = { version = "*", features = ["std", "alloc"] }
rand = { version = "0.8.5", features = ["getrandom"] }
rand_core = "0.6.3"
serde = "1.0.136"
serde_derive = "1.0.136"
serde_json = "1.0.79"
static-iref = "2.0.0"
tokio = { version = "1.17.0", features = ["time", "macros", "rt-multi-thread"] }
tracing = "0.1.32"
url = { version = "2.2.2", features = ["serde"] }
uuid = { version = "0.8.2", features = ["serde", "v4"] }

[build-dependencies]
json        = "0.12.4"
lazy_static = "1.4.0"

[dev-dependencies]
insta = { version = "1.14.0", features = ["redactions", "toml"] }
proptest = "1.0.0"

[features]
strict = []
