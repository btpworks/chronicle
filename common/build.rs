use std::{env, fs, path::PathBuf};

include!("./src/context.rs");

fn main() {
    let out_str = env::var("OUT_DIR").unwrap();
    let out_path = PathBuf::from(&out_str);
    let mut out_path = out_path
        .ancestors() // .../target/<debug|release>/build/example-<SHA>/out
        .skip(3) // .../target/<debug|release>
        .next()
        .unwrap()
        .to_owned();
    out_path.push("assets");

    if !out_path.exists() {
        fs::create_dir(&out_path).expect("Could not create assets dir");
    }

    let context = &PROV;

    std::fs::write(
        std::path::Path::new(&format!(
            "{}/context.json",
            out_path.as_os_str().to_string_lossy(),
        )),
        context.pretty(2),
    )
    .unwrap();
}
