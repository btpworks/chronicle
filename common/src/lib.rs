#[macro_use]
extern crate serde_derive;

pub mod ledger;
pub mod models;
pub mod signing;
