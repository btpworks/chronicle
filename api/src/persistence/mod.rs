use std::{cell::RefCell, collections::HashMap, str::FromStr};

use common::{
    models::{
        Activity, ActivityId, Agent, AgentId, ChronicleTransaction, Namespace, NamespaceId,
        ProvModel,
    },
    vocab::Chronicle,
};
use custom_error::custom_error;
use derivative::Derivative;
use diesel::{dsl::max, prelude::*, sqlite::SqliteConnection};
use diesel_migrations::{embed_migrations, EmbeddedMigrations, MigrationHarness};
use tracing::{instrument, trace};
use uuid::Uuid;

mod query;
mod schema;

pub const MIGRATIONS: EmbeddedMigrations = embed_migrations!();

custom_error! {pub StoreError
    Db{source: diesel::result::Error}                           = "Database operation failed",
    DbConnection{source: diesel::ConnectionError}               = "Database connection failed",
    DbMigration{source: diesel_migrations::MigrationError}      = "Database migration failed",
    Uuid{source: uuid::Error}                                   = "Invalid UUID string",
    RecordNotFound{}                                            = "Could not locate record in store",
    InvalidNamespace{}                                          = "Could not find namespace",
    ModelDoesNotContainActivity{activityid: ActivityId}         = "Could not locate {} in activities",
    ModelDoesNotContainAgent{agentid: AgentId}                  = "Could not locate {} in agents",
}

#[derive(Derivative)]
#[derivative(Debug)]
pub struct Store {
    #[derivative(Debug = "ignore")]
    connection: RefCell<SqliteConnection>,
}

impl Store {
    pub fn new(database_url: &str) -> Result<Self, StoreError> {
        let mut connection = SqliteConnection::establish(database_url)?;
        connection.run_pending_migrations(MIGRATIONS).unwrap();

        Ok(Store {
            connection: connection.into(),
        })
    }

    /// Apply a chronicle transaction to the store idempotently and return a prov model relevant to the transaction
    #[instrument]
    pub fn apply(&self, tx: &ChronicleTransaction) -> Result<ProvModel, StoreError> {
        let model = ProvModel::from_tx(vec![tx]);

        trace!(?model);

        self.idempotently_apply_model(&model)?;

        Ok(model)
    }

    fn idempotently_apply_model(&self, model: &ProvModel) -> Result<(), StoreError> {
        for (_, ns) in model.namespaces.iter() {
            self.apply_namespace(ns)?
        }
        for (_, agent) in model.agents.iter() {
            self.apply_agent(agent, &model.namespaces)?
        }
        for (_, activity) in model.activities.iter() {
            self.apply_activity(activity, &model.namespaces)?
        }

        for (activityid, agentid) in model.was_associated_with.iter() {
            for agentid in agentid {
                self.apply_was_associated_with(model, activityid, agentid)?;
            }
        }
        Ok(())
    }

    #[instrument]
    fn apply_namespace(
        &self,
        Namespace {
            ref name, ref uuid, ..
        }: &Namespace,
    ) -> Result<(), StoreError> {
        diesel::insert_or_ignore_into(schema::namespace::table)
            .values(&query::NewNamespace {
                name,
                uuid: &uuid.to_string(),
            })
            .execute(&mut *self.connection.borrow_mut())?;

        Ok(())
    }

    #[instrument]
    pub(crate) fn namespace_by_name(&self, namespace: &str) -> Result<NamespaceId, StoreError> {
        use self::schema::namespace::dsl as ns;
        let ns = ns::namespace
            .filter(ns::name.eq(namespace))
            .first::<query::Namespace>(&mut *self.connection.borrow_mut())
            .optional()?
            .ok_or(StoreError::RecordNotFound {})?;

        Ok(Chronicle::namespace(&ns.name, &Uuid::from_str(&ns.uuid)?).into())
    }

    #[instrument]
    fn apply_agent(
        &self,
        Agent {
            ref name,
            namespaceid,
            publickey,
            id: _,
        }: &Agent,
        ns: &HashMap<NamespaceId, Namespace>,
    ) -> Result<(), StoreError> {
        let namespace = ns.get(namespaceid).ok_or(StoreError::InvalidNamespace {})?;
        diesel::insert_or_ignore_into(schema::agent::table)
            .values(&query::NewAgent {
                name,
                namespace: &namespace.name,
                current: 0,
                publickey: publickey.as_deref(),
                privatekeypath: None,
            })
            .execute(&mut *self.connection.borrow_mut())?;

        Ok(())
    }

    #[instrument]
    fn apply_activity(
        &self,
        Activity {
            ref name,
            id,
            namespaceid,
            started,
            ended,
        }: &Activity,
        ns: &HashMap<NamespaceId, Namespace>,
    ) -> Result<(), StoreError> {
        let namespace = ns.get(namespaceid).ok_or(StoreError::InvalidNamespace {})?;
        diesel::insert_or_ignore_into(schema::activity::table)
            .values(&query::NewActivity {
                name,
                namespace: &namespace.name,
                started: started.map(|t| t.naive_utc()),
                ended: ended.map(|t| t.naive_utc()),
            })
            .execute(&mut *self.connection.borrow_mut())?;

        Ok(())
    }

    pub(crate) fn store_pk_path(
        &self,
        name: String,
        namespace: String,
        privatekeypath: String,
    ) -> Result<(), StoreError> {
        use schema::agent::dsl;
        diesel::update(
            schema::agent::table.filter(dsl::name.eq(name).and(dsl::namespace.eq(namespace))),
        )
        .set(dsl::privatekeypath.eq(Some(privatekeypath)))
        .execute(&mut *self.connection.borrow_mut())?;

        Ok(())
    }

    pub(crate) fn use_agent(&self, name: String, namespace: String) -> Result<(), StoreError> {
        use schema::agent::dsl;
        diesel::update(schema::agent::table.filter(dsl::current.ne(0)))
            .set(dsl::current.eq(0))
            .execute(&mut *self.connection.borrow_mut())?;

        diesel::update(
            schema::agent::table.filter(dsl::name.eq(name).and(dsl::namespace.eq(namespace))),
        )
        .set(dsl::current.eq(1))
        .execute(&mut *self.connection.borrow_mut())?;

        Ok(())
    }

    pub(crate) fn get_current_agent(&self) -> Result<query::Agent, StoreError> {
        use schema::agent::dsl;
        Ok(schema::agent::table
            .filter(dsl::current.ne(0))
            .first::<query::Agent>(&mut *self.connection.borrow_mut())?)
    }

    /// Ensure the name is unique within the namespace, if not, then postfix the rowid
    pub(crate) fn disambiguate_agent_name(&self, name: &str) -> Result<String, StoreError> {
        use schema::agent::dsl as agentdsl;

        let ambiguous = schema::agent::table
            .select(max(agentdsl::id))
            .first::<Option<i32>>(&mut *self.connection.borrow_mut())?;

        Ok(format!("{}_{}", name, ambiguous.unwrap_or_default()))
    }

    /// Ensure the name is unique within the namespace, if not, then postfix the rowid
    pub(crate) fn disambiguate_activity_name(&self, name: &str) -> Result<String, StoreError> {
        use schema::activity::dsl;

        let ambiguous = schema::activity::table
            .select(max(dsl::id))
            .first::<Option<i32>>(&mut *self.connection.borrow_mut())?;

        Ok(format!("{}_{}", name, ambiguous.unwrap_or_default()))
    }

    /// Fetch the activity record for the IRI
    fn activity_by_activity_name_and_namespace(
        &self,
        name: &str,
        namespace: &str,
    ) -> Result<query::Activity, StoreError> {
        use schema::activity::dsl;

        Ok(schema::activity::table
            .filter(dsl::name.eq(name).and(dsl::namespace.eq(namespace)))
            .first::<query::Activity>(&mut *self.connection.borrow_mut())?)
    }

    /// Fetch the agent record for the IRI
    fn agent_by_agent_name_and_namespace(
        &self,
        name: &str,
        namespace: &str,
    ) -> Result<query::Agent, StoreError> {
        use schema::agent::dsl;

        Ok(schema::agent::table
            .filter(dsl::name.eq(name).and(dsl::namespace.eq(namespace)))
            .first::<query::Agent>(&mut *self.connection.borrow_mut())?)
    }

    fn apply_was_associated_with(
        &self,
        model: &ProvModel,
        activityid: &common::models::ActivityId,
        agentid: &common::models::AgentId,
    ) -> Result<(), StoreError> {
        let provagent =
            model
                .agents
                .get(agentid)
                .ok_or_else(|| StoreError::ModelDoesNotContainAgent {
                    agentid: agentid.clone(),
                })?;
        let provactivity = model.activities.get(activityid).ok_or_else(|| {
            StoreError::ModelDoesNotContainActivity {
                activityid: activityid.clone(),
            }
        })?;

        let storedactivity = self.activity_by_activity_name_and_namespace(
            &provactivity.name,
            provactivity.namespaceid.decompose().0,
        )?;

        let storedagent = self.agent_by_agent_name_and_namespace(
            &provagent.name,
            provagent.namespaceid.decompose().0,
        )?;

        use schema::wasassociatedwith::dsl as link;
        diesel::insert_or_ignore_into(schema::wasassociatedwith::table)
            .values((
                &link::activity.eq(storedactivity.id),
                &link::agent.eq(storedagent.id),
            ))
            .execute(&mut *self.connection.borrow_mut())?;

        Ok(())
    }

    /// Get the named acitvity or the last started one, a useful context aware shortcut for the CLI
    pub(crate) fn get_activity_by_name_or_last_started(
        &self,
        name: Option<String>,
        namespace: Option<String>,
    ) -> Result<query::Activity, StoreError> {
        use schema::activity::dsl;

        match (name, namespace) {
            (Some(name), Some(namespace)) => {
                Ok(self.activity_by_activity_name_and_namespace(&name, &namespace)?)
            }
            _ => Ok(schema::activity::table
                .order(dsl::started)
                .first::<query::Activity>(&mut *self.connection.borrow_mut())?),
        }
    }
}
