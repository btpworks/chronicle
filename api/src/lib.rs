mod persistence;

use chrono::{DateTime, Utc};
use custom_error::*;
use derivative::*;

use persistence::Store;
use std::{path::Path, task::Waker};

use common::{
    ledger::{LedgerWriter, SubmissionError},
    models::{
        ChronicleTransaction, CreateActivity, CreateAgent, CreateNamespace, ProvModel, RegisterKey,
        StartActivity,
    },
    signing::SignerError,
    vocab::Chronicle as ChronicleVocab,
};

use tracing::{debug, instrument};

use user_error::UFE;
use uuid::Uuid;

custom_error! {pub ApiError
    Store{source: persistence::StoreError}                      = "Storage",
    Iri{source: iref::Error}                                    = "Invalid IRI",
    JsonLD{source: json_ld::Error}                              = "Json LD processing",
    Ledger{source: SubmissionError}                             = "Ledger error",
    Signing{source: SignerError}                                = "Signing",
    NamespaceNotFound{name: String}                             = "Namespace {} not found, please create it",
    NoCurrentAgent{}                                            = "No agent is currently in use, please call agent use",
}

impl UFE for ApiError {}

#[derive(Debug)]
pub enum NamespaceCommand {
    Create { name: String },
}

#[derive(Debug)]
pub enum AgentCommand {
    Create {
        name: String,
        namespace: String,
    },
    RegisterKey {
        name: String,
        namespace: String,
        public: String,
        private: Option<String>,
    },
    Use {
        name: String,
        namespace: String,
    },
}

#[derive(Debug)]
pub enum ActivityCommand {
    Create {
        name: String,
        namespace: String,
    },
    Start {
        name: String,
        namespace: String,
        time: Option<DateTime<Utc>>,
    },
    End {
        name: Option<String>,
        namespace: Option<String>,
    },
    Use {
        name: Option<String>,
        namespace: Option<String>,
    },
    Generate {
        name: Option<String>,
        namespace: Option<String>,
        entityname: String,
    },
    Sign {
        name: Option<String>,
        namespace: Option<String>,
        entityname: String,
    },
}

#[derive(Debug)]
pub enum ApiCommand {
    NameSpace(NamespaceCommand),
    Agent(AgentCommand),
    Activity(ActivityCommand),
}

#[derive(Debug)]
pub enum ApiResponse {
    Unit,
    Prov(ProvModel),
}

#[derive(Derivative)]
#[derivative(Debug)]
pub struct Api {
    #[derivative(Debug = "ignore")]
    ledger: Box<dyn LedgerWriter>,
    #[derivative(Debug = "ignore")]
    store: persistence::Store,
    #[derivative(Debug = "ignore")]
    uuidsource: Box<dyn Fn() -> Uuid>,
}

impl Api {
    #[instrument(skip(ledger, uuidgen))]
    pub fn new<F>(
        database_url: &str,
        ledger: Box<dyn LedgerWriter>,
        _secret_path: &Path,
        uuidgen: F,
    ) -> Result<Self, ApiError>
    where
        F: Fn() -> Uuid,
        F: 'static,
    {
        Ok(Api {
            ledger: ledger,
            store: Store::new(database_url)?,
            uuidsource: Box::new(uuidgen),
        })
    }

    /// Our resources all assume a namespace, or the default namspace, so automatically create it by name if it doesn't exist
    #[instrument]
    fn ensure_namespace(&self, namespace: &str) -> Result<(), ApiError> {
        let ns = self.store.namespace_by_name(namespace);

        if ns.is_err() {
            debug!(namespace, "Namespace does not exist, creating");
            self.create_namespace(namespace)?;
        }

        Ok(())
    }

    #[instrument]
    fn create_namespace(&self, name: &str) -> Result<ApiResponse, ApiError> {
        let uuid = (self.uuidsource)();
        let iri = ChronicleVocab::namespace(name, &uuid);

        let tx = ChronicleTransaction::CreateNamespace(CreateNamespace {
            id: iri.into(),
            name: name.to_owned(),
            uuid,
        });

        self.ledger.submit(vec![&tx])?;

        Ok(ApiResponse::Prov(self.store.apply(&tx)?))
    }

    #[instrument]
    fn create_agent(&self, name: &str, namespace: &str) -> Result<ApiResponse, ApiError> {
        let name = self.store.disambiguate_agent_name(name)?;
        self.ensure_namespace(namespace)?;

        let tx = ChronicleTransaction::CreateAgent(CreateAgent {
            name: name.to_owned(),
            id: ChronicleVocab::agent(&name).into(),
            namespace: self.store.namespace_by_name(namespace).map_err(|_| {
                ApiError::NamespaceNotFound {
                    name: namespace.to_owned(),
                }
            })?,
        });

        self.ledger.submit(vec![&tx])?;

        Ok(ApiResponse::Prov(self.store.apply(&tx)?))
    }

    #[instrument]
    pub fn dispatch(&self, command: ApiCommand) -> Result<ApiResponse, ApiError> {
        match command {
            ApiCommand::NameSpace(NamespaceCommand::Create { name }) => {
                self.create_namespace(&name)
            }
            ApiCommand::Agent(AgentCommand::Create { name, namespace }) => {
                self.create_agent(&name, &namespace)
            }
            ApiCommand::Agent(AgentCommand::RegisterKey {
                name,
                namespace,
                public,
                private,
            }) => self.register_key(name, namespace, public, private),
            ApiCommand::Agent(AgentCommand::Use { name, namespace }) => {
                self.use_agent(name, namespace)
            }
            ApiCommand::Activity(ActivityCommand::Create { name, namespace }) => {
                self.create_activity(name, namespace)
            }
            ApiCommand::Activity(ActivityCommand::Start {
                name,
                namespace,
                time,
            }) => self.start_activity(name, namespace, time),
            _ => todo!(),
        }
    }

    #[instrument]
    fn register_key(
        &self,
        name: String,
        namespace: String,
        publickey: String,
        privatekeypath: Option<String>,
    ) -> Result<ApiResponse, ApiError> {
        self.ensure_namespace(&namespace)?;
        let namespaceid = self.store.namespace_by_name(&namespace)?;

        let tx = ChronicleTransaction::RegisterKey(RegisterKey {
            id: ChronicleVocab::agent(&name).into(),
            name: name.clone(),
            namespace: namespaceid,
            publickey,
        });

        self.ledger.submit(vec![&tx])?;
        self.store.apply(&tx)?;
        privatekeypath.map(|pk| self.store.store_pk_path(name, namespace, pk));

        Ok(ApiResponse::Prov(self.store.apply(&tx)?))
    }

    fn use_agent(&self, name: String, namespace: String) -> Result<ApiResponse, ApiError> {
        self.store.use_agent(name, namespace)?;

        Ok(ApiResponse::Unit)
    }

    #[instrument]
    fn create_activity(&self, name: String, namespace: String) -> Result<ApiResponse, ApiError> {
        self.ensure_namespace(&namespace)?;
        let name = self.store.disambiguate_activity_name(&name)?;
        let namespace = self.store.namespace_by_name(&namespace)?;
        let id = ChronicleVocab::activity(&name);
        let tx = ChronicleTransaction::CreateActivity(CreateActivity {
            namespace,
            id: id.into(),
            name,
        });

        self.ledger.submit(vec![&tx])?;

        Ok(ApiResponse::Prov(self.store.apply(&tx)?))
    }

    pub(crate) fn start_activity(
        &self,
        name: String,
        namespace: String,
        time: Option<DateTime<Utc>>,
    ) -> Result<ApiResponse, ApiError> {
        let agent = self
            .store
            .get_current_agent()
            .map_err(|_| ApiError::NoCurrentAgent {})?;

        let name = self.store.disambiguate_activity_name(&name)?;
        let namespace = self.store.namespace_by_name(&namespace)?;
        let id = ChronicleVocab::activity(&name);
        let tx = ChronicleTransaction::StartActivity(StartActivity {
            namespace,
            id: id.into(),
            agent: ChronicleVocab::agent(&agent.name).into(),
            time: time.unwrap_or(Utc::now()),
        });

        self.ledger.submit(vec![&tx])?;

        Ok(ApiResponse::Prov(self.store.apply(&tx)?))
    }
}

#[cfg(test)]
mod test {
    use chrono::{TimeZone, Utc};
    use common::{ledger::InMemLedger, models::Activity};
    use tempfile::TempDir;
    use tracing::Level;
    use uuid::Uuid;

    use crate::{ActivityCommand, AgentCommand, Api, ApiCommand, ApiResponse, NamespaceCommand};

    fn test_api() -> Api {
        tracing_subscriber::fmt()
            .pretty()
            .with_max_level(Level::TRACE)
            .try_init()
            .ok();

        let secretpath = TempDir::new().unwrap();
        Api::new(
            "file::memory:",
            Box::new(InMemLedger::default()),
            &secretpath.into_path(),
            || Uuid::parse_str("5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea").unwrap(),
        )
        .unwrap()
    }

    #[test]
    fn create_namespace() {
        let prov = test_api()
            .dispatch(ApiCommand::NameSpace(NamespaceCommand::Create {
                name: "testns".to_owned(),
            }))
            .unwrap();

        match prov {
            ApiResponse::Prov(prov) => {
                insta::assert_snapshot!(prov.to_json().0.pretty(3))
            }
            _ => unreachable!(),
        }
    }

    #[test]
    fn create_agent() {
        let api = test_api();

        api.dispatch(ApiCommand::NameSpace(NamespaceCommand::Create {
            name: "testns".to_owned(),
        }))
        .unwrap();

        let prov = api
            .dispatch(ApiCommand::Agent(AgentCommand::Create {
                name: "testagent".to_owned(),
                namespace: "testns".to_owned(),
            }))
            .unwrap();

        match prov {
            ApiResponse::Prov(prov) => {
                insta::assert_snapshot!(prov.to_json().0.pretty(3))
            }
            _ => unreachable!(),
        }
    }

    #[test]
    fn agent_publiv_key() {
        let api = test_api();

        api.dispatch(ApiCommand::NameSpace(NamespaceCommand::Create {
            name: "testns".to_owned(),
        }))
        .unwrap();

        let prov = api
            .dispatch(ApiCommand::Agent(AgentCommand::RegisterKey {
                name: "testagent".to_owned(),
                namespace: "testns".to_owned(),
                public: "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA".to_owned(),
                private: Some("/test/path".to_owned()),
            }))
            .unwrap();

        match prov {
            ApiResponse::Prov(prov) => {
                insta::assert_snapshot!(prov.to_json().0.pretty(3))
            }
            _ => unreachable!(),
        }
    }

    #[test]
    fn create_activity() {
        let api = test_api();

        let prov = api
            .dispatch(ApiCommand::Activity(ActivityCommand::Create {
                name: "testactivity".to_owned(),
                namespace: "testns".to_owned(),
            }))
            .unwrap();

        match prov {
            ApiResponse::Prov(prov) => {
                insta::assert_snapshot!(prov.to_json().0.pretty(3))
            }
            _ => unreachable!(),
        }
    }

    #[test]
    fn start_activity() {
        let api = test_api();

        api.dispatch(ApiCommand::Agent(AgentCommand::Create {
            name: "testagent".to_owned(),
            namespace: "testns".to_owned(),
        }))
        .unwrap();

        api.dispatch(ApiCommand::Agent(AgentCommand::Use {
            name: "testagent_0".to_owned(),
            namespace: "testns".to_owned(),
        }))
        .unwrap();

        let prov = api
            .dispatch(ApiCommand::Activity(ActivityCommand::Start {
                name: "testactivity".to_owned(),
                namespace: "testns".to_owned(),
                time: Some(Utc.ymd(2014, 7, 8).and_hms(9, 10, 11)),
            }))
            .unwrap();

        match prov {
            ApiResponse::Prov(prov) => {
                insta::assert_snapshot!(prov.to_json().0.pretty(3))
            }
            _ => unreachable!(),
        }
    }
}
