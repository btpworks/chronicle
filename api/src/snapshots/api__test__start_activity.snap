---
source: api/src/lib.rs
expression: prov.to_json().0.pretty(3)

---
[
   {
      "@id": "http://blockchaintp.com/chonicle/ns#:ns:testns:5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea",
      "@type": "http://blockchaintp.com/chonicle/ns#Namespace",
      "http://www.w3.org/2000/01/rdf-schema#label": [
         {
            "@value": "testns"
         }
      ]
   },
   {
      "@id": "http://blockchaintp.com/chonicle/ns#:agent:testagent_0",
      "@type": "http://www.w3.org/ns/prov#Agent",
      "http://www.w3.org/2000/01/rdf-schema#label": [
         {
            "@value": "agent"
         }
      ],
      "http://blockchaintp.com/chonicle/ns#hasNamespace": [
         {
            "@id": "http://blockchaintp.com/chonicle/ns#:ns:testns:5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea"
         }
      ]
   },
   {
      "@id": "http://blockchaintp.com/chonicle/ns#:activity:testactivity_0",
      "@type": "http://www.w3.org/ns/prov#Activity",
      "http://www.w3.org/2000/01/rdf-schema#label": [
         {
            "@value": "activity"
         }
      ],
      "http://www.w3.org/ns/prov#startedAtTime": [
         {
            "@value": "2014-07-08T09:10:11+00:00"
         }
      ],
      "http://www.w3.org/ns/prov#wasAssociatedWith": [
         {
            "@id": "http://blockchaintp.com/chonicle/ns#:agent:testagent_0"
         }
      ],
      "http://blockchaintp.com/chonicle/ns#hasNamespace": [
         {
            "@id": "http://blockchaintp.com/chonicle/ns#:ns:testns:5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea"
         }
      ]
   }
]
